src/storage/src/sink/kafka.rs

```diff
@@ -89,1152 +138,504 @@ where
+                // leaves the rest of the objects in the cluster unaffected.
+                error!("{err}");
+                return Err(anyhow!("{err}").into());
+            }
 
-                            s.send(record).await;
-                            sink_statistics.inc_messages_staged_by(1);
-                            sink_statistics.inc_bytes_staged_by(size_for_stats);
+            info!(
+                "{name}: as_of: {}, resume upper: {}",
+                as_of.pretty(),
+                resume_upper.pretty()
+            );
 
-                            // advance to the next repetition of this row, or the next row if all
-                            // repetitions are exhausted
-                            repeat_counter += 1;
-                            if repeat_counter == encoded_row.count {
-                                repeat_counter = 0;
-                                s.metrics.rows_queued.dec();
+            // The section below relies on TotalOrder for correctness so we'll work with timestamps
+            // directly to make sure this doesn't compile if someone attempts to make this operator
+            // generic over partial orders in the future.
+            let Some(mut upper) = resume_upper.clone().into_option() else {
+                return Ok(());
+            };
+            let mut deferred_updates = vec![];
+            let mut extra_updates = vec![];
+            // We must wait until we have data to commit before starting a transaction because
+            // Kafka doesn't have a heartbeating mechanism to keep a transaction open indefinitely.
+            // This flag tracks whether we have started the transaction.
+            let mut transaction_begun = false;
+            while let Some(event) = input.next_mut().await {
+                match event {
+                    Event::Data(_cap, batch) => {
+                        for ((key, value), time, diff) in batch.drain(..) {
+                            // We want to publish updates in time order and we know that we have
+                            // already committed all times not beyond `upper`. Therefore, if this
+                            // update happens *exactly* at upper then it is the minimum pending
+                            // time and so emitting it now will not violate the timestamp publish
+                            // order. This optimization is load bearing because it is the mechanism
+                            // by which we incrementally stream the initial snapshot out to Kafka
+                            // instead of buffering it all in memory first. This argument doesn't
+                            // hold for partially ordered time because many different timestamps
+                            // can be *exactly* at upper but we can't know ahead of time which one
+                            // will be advanced in the next progress message.
```

Comment:
this optimization is quite subtle, is there ci that shows it necessary?

Reply:
No but there was a recent issue with a customer where a sink was in an OOM loop because it was repeatedly trying to sink the snapshot. https://materializeinc.slack.com/archives/C01CFKM1QRF/p1701728894506549?thread_ts=1701726334.717919&cid=C01CFKM1QRF